lib/puppet/transaction.rb

```diff
@@ -338,17 +353,30 @@ def resources_by_provider(type_name, provider_name)
   # types, just providers.
   def prefetch(provider_class, resources)
     type_name = provider_class.resource_type.name
-    return if @prefetched_providers[type_name][provider_class.name]
+    return if @prefetched_providers[type_name][provider_class.name] or
+      @prefetch_failed_providers[type_name][provider_class.name]
     Puppet.debug "Prefetching #{provider_class.name} resources for #{type_name}"
     begin
       provider_class.prefetch(resources)
-    rescue LoadError, Puppet::MissingCommand => detail
+    rescue Exception => detail
+      message = _("Could not prefetch %{type_name} provider '%{name}': %{detail}") % { type_name: type_name, name: provider_class.name, detail: detail }
+      if(!detail.is_a?(LoadError) && !detail.is_a?(Puppet::MissingCommand))
+        raise unless Puppet.settings[:future_features]
+        
+        @prefetch_failed_providers[type_name][provider_class.name] = true
+
+        dummy_until_i_figure_out_the_mocha_thing
+      end
       #TRANSLATORS `prefetch` is a function name and should not be translated
-      Puppet.log_exception(detail, _("Could not prefetch %{type_name} provider '%{name}': %{detail}") % { type_name: type_name, name: provider_class.name, detail: detail })
+      Puppet.log_exception(detail, message)
     end
     @prefetched_providers[type_name][provider_class.name] = true
   end
 
+  #This is a dummy method to help with spec tests until I can work out the right Mocha incantation to replace it.
+  def dummy_until_i_figure_out_the_mocha_thing()
```

Comment:
Do we still needs this? And if so can you give a hit as to why?

Reply:
This should definitely go away.